### PR TITLE
Extend support of allowedFields to getMatchingFieldNames and getAllFields

### PR DESCRIPTION
**Pull request by javanna**
The SearchExecutionContext supports the notion of allowed fields, provided via a specific setter method. Fields are though only filtered for the getFieldType method. There needs to be consistency between getMatchingFieldNames and getFieldType. In fact there are places in the code where getMatchingFieldNames is called to resolve field name patterns, and later getFieldType is called on each of the resolved fields. If the former resolves to one field that we can't retrieve a field type for, that is unexpected and to be considered a bug.

In addition, this commit adds consistency for getAllFields: this is only called by field caps, hence a different codepath that does not seem to set allowed fields for now, but it's important for the context to provide consistency around fields access, especially for methods that are as broad as getAllFields, despite their currently very specific usage.

This surfaced as we are trying to move fetching of the `_ignored` field to use value fetchers, which use a search execution context and resolve the field type, whereas until now they are retrieved directly via StoredFieldsPhase and completely bypass such check.

This commit also adds a test that was missing around verifying that SearchExecutionContext applies the allowedFields predicate when provided.

### CODE REVIEW

**Comment by salvatore-campagna** [comment]
LGTM...I just left a comment.

**Review by tteofili — APPROVED** [review]
the changes look good, Luca ;)

File: server/src/main/java/org/elasticsearch/index/query/QueryRewriteContext.java [hunk 1]
```diff
@@ -318,35 +319,45 @@ public boolean indexMatches(String pattern) {
...
+                for (String name : runtimeMappings.keySet()) {
+                    if (Regex.simpleMatch(pattern, name)) {
+                        matches.add(name);
+                    }
                 }
             }
         }
-        return matches;
+        // If the field is not allowed, behave as if it is not mapped
+        return allowedFields == null ? matches : matches.stream().filter(allowedFields).collect(Collectors.toSet());
```
Comment (salvatore-campagna) [inline server/src/main/java/org/elasticsearch/index/query/QueryRewriteContext.java:343]: If I understand correctly this is just the same we were doing before + an additional filter on allowed fields based n field names...maybe we can just extract the existing method into something like a private `getMatchingFieldNamesInternal` to which we apply the filter?

**Review by salvatore-campagna — APPROVED** [review]

File: server/src/main/java/org/elasticsearch/index/query/QueryRewriteContext.java (hunk 1 above)
[the remark this answers is not in the document]
Comment (javanna) [reply server/src/main/java/org/elasticsearch/index/query/QueryRewriteContext.java:343]: I am not too sure what is best here. Skip the filtering entirely when the predicate is null, or always filter although the predicate always returns true? I went for the former, which leaves things as close as they are outside of api keys queries (when setAllowedFields is not called).

Comment (javanna) [reply server/src/main/java/org/elasticsearch/index/query/QueryRewriteContext.java:343]: what is the gain compared to a single method?

[the remark this answers is not in the document]
Comment (salvatore-campagna) [reply server/src/main/java/org/elasticsearch/index/query/QueryRewriteContext.java:343]: I agree that, maybe, performance-wise not applying the filter at all might be better if there is nothing to filter out...which is the reason why I said it looks good to me...it is just that all those null checks are annoying for me.

**Pull request merged**